File: client-android/lib/src/main/kotlin/stasis/client_android/lib/discovery/ServiceApiEndpoint.kt
package stasis.client_android.lib.discovery

import stasis.client_android.lib.model.core.networking.EndpointAddress

sealed class ServiceApiEndpoint {
    abstract val id: String

    data class Api(val uri: String) : ServiceApiEndpoint() {
        override val id: String by lazy { "api__$uri" }
    }

    data class Core(val address: EndpointAddress) : ServiceApiEndpoint() {
        override val id: String by lazy {
            when (address) {
                is EndpointAddress.HttpEndpointAddress -> "core_http__${address.uri}"
                is EndpointAddress.GrpcEndpointAddress -> "core_grpc__${address.host}:${address.port}"
            }
        }

    }

    data class Discovery(val uri: String) : ServiceApiEndpoint() {
        override val id: String by lazy { "discovery__$uri" }
    }
}

File: core/src/main/scala/stasis/core/discovery/ServiceApiEndpoint.scala
package stasis.core.discovery

import stasis.core.networking.EndpointAddress
import stasis.core.networking.grpc.GrpcEndpointAddress
import stasis.core.networking.http.HttpEndpointAddress

sealed trait ServiceApiEndpoint {
  def id: String
}

object ServiceApiEndpoint {
  final case class Api(uri: String) extends ServiceApiEndpoint {
    override lazy val id: String = s"api__$uri"
  }

  final case class Core(address: EndpointAddress) extends ServiceApiEndpoint {
    @SuppressWarnings(Array("org.wartremover.warts.Throw"))
    override lazy val id: String = address match {
      case HttpEndpointAddress(uri)           => s"core_http__${uri.toString}"
      case GrpcEndpointAddress(host, port, _) => s"core_grpc__$host:${port.toString}"
      case other => throw new IllegalArgumentException(s"Unexpected address provided: [${other.getClass.getSimpleName}]")
    }
  }

  final case class Discovery(uri: String) extends ServiceApiEndpoint {
    override lazy val id: String = s"discovery__$uri"
  }
}
